### PR TITLE
Update Lightroom to 6.3

### PR DESCRIPTION
**Pull request by faun**
This is a complicated update from the 5.x branch to the 6.x branch. There's a requirement that 6.x be installed and then the updater must be run on top of that. And of course, in typical Adobe style, you have to run two installers that then ask for an Adobe ID _and_ a serial number. Let me know if I should move the base installer to the versions repository as `adobe-photoshop-lightroom-6` or rename it somehow. I'm open to suggestions.

TODO:
- [x] Move 6.0 installer to hombrew-versions
- [x] Make 6.3 install rely on 6.0 from hombrew-versions
- [x] Move 5.7.1 installer to homebrew-versions instead of clobbering it here
- [x] Ensure uninstalled application files match what's installed by 6.x 
- [x] Convert 6.0 to use silent, automated installer
- [ ] Convert 6.3 patch installer to use silent, automated installer

### CODE REVIEW

File: Casks/adobe-photoshop-lightroom.rb [hunk 1]
```diff
@@ -1,13 +1,14 @@
 cask :v1 => 'adobe-photoshop-lightroom' do
-  version '5.7.1'
-  sha256 '155a91e2c90927a05ccaa244a99fed4784fa7cf26d08c634f5f111629f6b0418'
+  version '6.3'
+  sha256 'f1282896be452a8515f92f5dd104695f54db114d5d06ba8d26175654cb26a524'
+
+  url "http://swupdl.adobe.com/updates/oobe/aam20/mac/AdobeLightroom-#{version.to_i}.0/#{version}/setup.dmg"
```
Comment (vitorgalvao) [inline Casks/adobe-photoshop-lightroom.rb]: This empty line shouldn’t be here.

**Comment by vitorgalvao** [comment]
> I'm open to suggestions.

I suggest everyone stops using Adobe products until they either go out of business or stop being user-hostile. Personally, I suggest [Affinity](https://affinity.serif.com/en-us/) as alternatives to (so far) Illustrator and Photoshop.

Seriously, though<sup>1</sup>, lets ask @winkelsdorf since he has done great work with Adobe installers.

---

<sup>1</sup> I was being dead serious, actually.

**Comment by winkelsdorf** [comment]
> I suggest everyone stops using Adobe products until they either go out of business or stop being user-hostile. Personally, I suggest Affinity as alternatives to (so far) Illustrator and Photoshop.

Haha :+1: Beside Adobe I'm using Affinity Photo (previously also Pixelmator), too. Yes, this is a total mess. I can't help but get the feeling that there is a department at Adobe solely responsible to make Users' life as difficult as possible.

> Seriously, though1, lets ask @winkelsdorf since he has done great work with Adobe installers.

I'll have a look at this.

**Comment by winkelsdorf** [comment]
What a mess, I don't understand why Adobe doesn't simply provide updated full installers!

Anyway, I would suggest keeping LR 5 and move all this to cask/Versions.

Regarding the naming convention: We should rename all existing LR casks to something following the convention of ```adobe-photoshop-lightroom-[version]```:
* ```cask/adobe-photoshop-lightroom.rb``` to ```versions/adobe-photoshop-lightroom-5.rb``` (move repo & rename)
* ```versions/lightroom4.rb```to ```versions/adobe-photoshop-lightroom-4.rb``` (rename)
* ```versions/lighroom3.rb``` to ```versions/adobe-photoshop-lightroom-3.rb``` (rename)

My current draft for the LR 6.0 Install (without the Updates) would be named:
```versions/adobe-photoshop-lightroom-600.rb ```

The update accordingly:
```versions/adobe-photoshop-lightroom-631.rb ```

That follows the naming convention when having fixed version numbers inside. The latter cask can also be named ```adobe-photoshop-lightroom-6-patches.rb``` (allowing later changes without breaking installer dotfiles). I am open to suggestions.

When a User tries to install 631 without having 600 installed, Cask will issue a warning because of the ```depends-on```.

My current draft for Installing 6.00, not tested yet as I need to setup an OSX VM for this:
```
cask :v1 => 'adobe-photoshop-lightroom-600' do
  version '6.0'
  sha256 '5c36e5fa76b8676144c4bba9790fe4c597daf350b2195a2088346b097f46a95f'

  # For correct download URL see links provided at
  # https://helpx.adobe.com/lightroom/kb/lightroom-downloads.html

  url "http://trials3.adobe.com/AdobeProducts/LTRM/#{version.to_i}/osx10/Lightroom_#{version.to_i}_LS11.dmg",
    :user_agent => :fake,
    :cookies => { 'MM_TRIALS' => '1234' }
  name 'Adobe Photoshop Lightroom 6.0'
  homepage 'http://www.adobe.com/de/products/photoshop-lightroom.html'
  license :commercial

  # staged_path not available in Installer/Uninstall Stanza, workaround by nesting with preflight/postflight
  # see https://github.com/caskroom/homebrew-cask/pull/8887
  # and https://github.com/caskroom/homebrew-versions/pull/296

  preflight do
    system '/usr/bin/killall', '-kill', 'SafariNotificationAgent'
    system '/usr/bin/sudo', '-E', '--', "#{staged_path}/#{version.to_i}/Install.app/Contents/MacOS/Install", '--mode=silent', "--deploymentFile=#{staged_path}/#{version.to_i}/deploy/AdobeLightroom6.install.xml"
  end

  uninstall_preflight do
    system '/usr/bin/killall', '-kill', 'SafariNotificationAgent'
    system '/usr/bin/sudo', '-E', '--', "#{staged_path}/#{version.to_i}/Install.app/Contents/MacOS/Install", "--mode=silent", "--deploymentFile=#{staged_path}/#{version.to_i}/deploy/AdobeLightroom6.remove.xml"
  #end

  caveats 'Installation or Uninstallation may fail with Exit Code 19 (Conflicting Processes running) if Browsers, Safari Notification Service or SIMBL Services are running or Adobe Creative Cloud or any other Adobe Products are already installed. See Logs in /Library/Logs/Adobe/Installers if Installation or Uninstallation fails, to identifify the conflicting processes.'
end
```

At least the Silent Install for LR 6.0 seems to be supported. Need to figure out, if the Patch Installer allows the same.

**Comment by vitorgalvao** [comment]
Thank you both for looking into it. Some notes, based on how homebrew-cask operates:

+ Only one version should be in this main repo, and it must be named without a version number.
+ Alternative versions should lose the `-` before the number (i.e. adobe-photoshop-lightroom-600` becomes `adobe-photoshop-lightroom600`). This is the way the other casks do it, and it’s a limitation/feature/requirement of how we handle file naming.

So based on the feedback, looks like this is what we should do:

+ Get rid of `adobe-photoshop-lightroom-base` as a name (too general, and may cause breakage in the future, when a base for 7 is needed), and have it instead as `adobe-photoshop-lightroom600` (or whatever, as long as it has a consistent, versioned name).
+ Add `adobe-photoshop-lightroom600` to [caskroom/versions](https://github.com/caskroom/homebrew-versions).
+ Update `adobe-photoshop-lightroom` here (to `6.3`, it seems), but still keep it named `adobe-photoshop-lightroom`.
+ Add `depends_on :cask => 'caskroom/versions/adobe-photoshop-lightroom600` to `adobe-photoshop-lightroom`.

**Comment by faun** [comment]
@vitorgalvao @winkelsdorf I have updated this PR and created https://github.com/caskroom/homebrew-versions/issues/1442. Let me know if I missed anything.

File: Casks/adobe-photoshop-lightroom.rb [hunk 2]
```diff
@@ -1,13 +1,13 @@
 cask :v1 => 'adobe-photoshop-lightroom' do
-  version '5.7.1'
-  sha256 '155a91e2c90927a05ccaa244a99fed4784fa7cf26d08c634f5f111629f6b0418'
+  version '6.3'
+  sha256 'f1282896be452a8515f92f5dd104695f54db114d5d06ba8d26175654cb26a524'
+  url "http://swupdl.adobe.com/updates/oobe/aam20/mac/AdobeLightroom-#{version.to_i}.0/#{version}/setup.dmg"
```
Comment (vitorgalvao) [inline Casks/adobe-photoshop-lightroom.rb]: There should be an empty line above `url` and none below.

File: Casks/adobe-photoshop-lightroom.rb [hunk 3]
```diff
@@ -16,4 +16,5 @@
                         '~/Library/Application Support/Adobe/Lightroom',
                         "~/Library/Preferences/com.adobe.Lightroom#{version.to_i}.plist",
                        ]
+  depends_on :cask => 'caskroom/versions/adobe-photoshop-lightroom600'
```
Comment (vitorgalvao) [inline Casks/adobe-photoshop-lightroom.rb]: Please add an empty line above this one.

**Comment by winkelsdorf** [comment]
> Thank you both for looking into it. Some notes, based on how homebrew-cask operates:

You're welcome!

> This is the way the other casks do it, and it’s a limitation/feature/requirement of how we handle file naming.

I am perfectly fine with. Just had no time to have a deeper look at the naming convention during work. I decided quickly upon what I saw in the Versions folder (btw somehow inconsistent at some places).

> So based on the feedback, looks like this is what we should do:
> 
>  Get rid of adobe-photoshop-lightroom-base as a name (too general, and may cause breakage in the future, when a base for 7 is needed) and have it instead as adobe-photoshop-lightroom600 (or whatever, as long as it has a consistent, versioned name).

Agreed.

@vitorgalvao & @faun Looks already good to me. Is the patch installer for 6.3 already silent and automated, too? If not (yet) you can merge this now and I'll try to make it silent and update the 2nd cask later today.

But anyway, please keep the previous 5.7.1 cask and move it as backup to Versions, naming it adobe-photoshop-lightroom571.rb.

**Comment by faun** [comment]
@winkelsdorf The patch installer for 6.3 is not silent and automated. It asks you to manually open the installer, just like the 6.0 installer.

**Comment by faun** [comment]
I can look into how to automate the installer later.

**Comment by faun** [comment]
I updated the description of this PR to clarify current progress.

**Comment by winkelsdorf** [comment]
> @winkelsdorf The patch installer for 6.3 is not silent and automated. It asks you to manually open the installer, just like the 6.0 installer.

Ok, thanks for checking back. Given the current PR https://github.com/caskroom/homebrew-versions/pull/1442/files#r46337105 the base should now already be silent.

> I can look into how to automate the installer later.

Whoever is quicker. But I can take care of, really. No need to do it all on your own.

**Comment by vitorgalvao** [comment]
> saw in the Versions folder (btw somehow inconsistent at some places)

Yeah, that whole repo is a mess. We should really have some rules to prevent so many things from entering, and some rules to make stuff leave when no longer useful. I have some ideas, but haven’t written them yet.

**Comment by winkelsdorf** [comment]
> Yeah, that whole repo is a mess. We should really have some rules to prevent so many things from entering, and some rules to make stuff leave when no longer useful. I have some ideas, but haven’t written them yet.

:+1: for the Rules. Probably a little parser is needed which checks all repo file names for ```[string]-[int].rb``` or something like that, and flag the casks. The same for the URLs in all Casks, which would help to see which casks need updating/can be removed.

**Comment by winkelsdorf** [comment]
@faun Here is my current draft for the 6.3 patch install (new "adobe-photoshop-lightroom.rb"):

Regarding your checklist this should be all.

```
cask :v1 => 'adobe-photoshop-lightroom' do
  version '6.3'
  sha256 'f1282896be452a8515f92f5dd104695f54db114d5d06ba8d26175654cb26a524'
  url "http://swupdl.adobe.com/updates/oobe/aam20/mac/AdobeLightroom-#{version.to_i}.0/#{version}/setup.dmg"

  name 'Adobe Photoshop Lightroom'
  homepage 'https://www.adobe.com/products/photoshop-lightroom.html'
  license :commercial

  uninstall :delete => "/Applications/Adobe Lightroom/Adobe Lightroom.app"
  zap       :delete => [
                        '~/Library/Application Support/Adobe/Lightroom',
                        "~/Library/Preferences/com.adobe.Lightroom#{version.to_i}.plist"
                       ]

  depends_on :cask => 'adobe-photoshop-lightroom600'

  # staged_path not available in Installer/Uninstall Stanza, workaround by nesting with preflight/postflight
  # see https://github.com/caskroom/homebrew-cask/pull/8887
  # and https://github.com/caskroom/homebrew-versions/pull/296

  preflight do
    system '/usr/bin/killall', '-kill', 'SafariNotificationAgent'
    system '/usr/bin/sudo', '-E', '--', "#{staged_path}/AdobePatchInstaller.app/Contents/MacOS/AdobePatchInstaller", '--mode=silent'
  end

#  uninstall_preflight do
#    system '/usr/bin/killall', '-kill', 'SafariNotificationAgent'
#    system '/usr/bin/sudo', '-E', '--', "#{staged_path}/AdobePatchInstaller.app/Contents/MacOS/AdobePatchInstaller", "--mode=silent"
#  end

  caveats 'Installation or Uninstallation may fail with Exit Code 19 (Conflicting Processes running) if Browsers, Safari Notification Service or SIMBL Services are running or Adobe Creative Cloud or any other Adobe Products are already installed. See Logs in /Library/Logs/Adobe/Installers if Installation or Uninstallation fails, to identify the conflicting processes.'

end
```

I can confirm that it works silently, but there is one issue: You cannot "Uninstall" a patch applied. 

@vitorgalvao Is there a way to invoke the Uninstallation of the depending cask? So that if a User tries to uninstall patch "6.3" the adobe-photoshop-lightroom600.rb is uninstalled, too?

Right now Uninstallation leaves a mess as it automatically removes parts of the App but without running Adobe Lightroom's uninstaller not everything is removed.

Or - probably better idea: Can the Uninstall of a Cask be aborted by Code? Let us show a message that the Patch cannot be uninstalled directly and the User must manually invoke uninstallation of depending cask "adobe-photoshop-lightroom600.rb" (probably by printing out depends_on value somehow)?

**Comment by vitorgalvao** [comment]
> Is there a way to invoke the Uninstallation of the depending cask?

Since we’re needing to use `uninstall_preflight` anyway having to resort to `system` calls, simplest way would be to simply call `brew cask uninstall adobe-photoshop-lightroom600`.

Simple to do, clear to understand when reading it, and should work. Considering all the hack jobs we need to have to support Adobe installers, that’s not even the worst.

**Comment by faun** [comment]
@vitorgalvao Is this what you're suggesting? https://github.com/faun/homebrew-cask/commit/0c507b687dd675efe6325d25d0e79a32c24a4d02#diff-901e4801959b94c53d6160f813cf08b6R30

**Comment by vitorgalvao** [comment]
> Is this what you're suggesting? https://github.com/faun/homebrew-cask/commit/0c507b687dd675efe6325d25d0e79a32c24a4d02#diff-901e4801959b94c53d6160f813cf08b6R30

Yes, but I was thinking we keep the uninstaller for this and *add* that line. Not sure if it’s needed, though, you two will know better.

**Comment by winkelsdorf** [comment]
@faun @vitorgalvao Looks good, I would just say let's remove the Uninstall section completely. So we just have zap and pre/postflights.

Uninstall is not needed here as vitorgalvao already noticed (and again wrong due to trying to uninstall a pkg, wrong plist name etc): The application and plists are already removed by the base cask.

I'll test the cask soon and let you know if it works :+1:

**Comment by winkelsdorf** [comment]
Finished my tests: Install & Uninstall of LR 6.3 works like a charm now. Good work :+1: 

Regarding my test: as the casks are not 'live' yet, I needed to adjust the depends_on path to fit my VM and use a local cask file. I am pretty sure that using depends_on with caskroom/versions is correct as it is now - it allows using the version base cask even if the user didn't tap versions. But don't nail me down on this - that's just a simple test I'd done.

**Comment by vitorgalvao** [comment]
Lets merge this! We got it done quicker than I expected. Thank you!

**Comment by winkelsdorf** [comment]
You're welcome! Thanks for merging and for asking me from start so these new casks are consistent with the Adobe Collection Casks :+1:

**Pull request merged**